lib/lightning/accounts/user.ex

```diff
@@ -164,14 +164,74 @@ defmodule Lightning.Accounts.User do
     |> put_change(:role, :superuser)
   end
 
-  def validate_email(changeset) do
+  @doc """
+  Validates that a user has access to all projects associated with a credential.
+  Returns a changeset with any project access errors.
+  """
+  def validate_project_access(changeset, credential) do
+    case get_field(changeset, :email) do
+      nil ->
+        changeset
+
+      email ->
+        case Lightning.Repo.exists?(User |> where(email: ^email)) do
+          false ->
+            changeset
+
+          true ->
+            user = Lightning.Accounts.get_user_by_email(email)
+
+            projects =
+              credential
+              |> Lightning.Credentials.diff_project_credentials_and_project_users(
+                user
+              )
+              |> Lightning.Projects.get_projects_by_ids()
+
+            if Enum.empty?(projects) do
+              changeset
+            else
+              projects_names = Enum.map_join(projects, ", ", & &1.name)
+
+              add_error(
+                changeset,
+                :email,
+                "User doesn't have access to these projects: #{projects_names}"
+              )
+            end
+        end
+    end
+  end
+
+  def validate_not_same_user(changeset, field, current_user, opts \\ []) do
+    message = opts[:message] || "Cannot be the same as current user"
```

Comment:
maybe reword `current user` to `the user logged in`

Reply:
Thanks, changed it to this

```elixir
recipient_email == sender_email ->
        Ecto.Changeset.add_error(
          changeset,
          :email,
          "You cannot transfer a credential to yourself"
        )
```